sgx-tools/dcap_linux.go

```diff
@@ -0,0 +1,166 @@
+package main // import "github.com/inclavare-containers/sgx-tools"
+
+import (
+	"bytes"
+	"encoding/binary"
+	"encoding/hex"
+	"fmt"
+	"github.com/go-restruct/restruct"
+	"github.com/golang/protobuf/proto"
+	"github.com/inclavare-containers/rune/libenclave/intelsgx"
+	pb "github.com/inclavare-containers/sgx-tools/proto"
+	"github.com/sirupsen/logrus"
+	"net"
+)
+
+const (
+	aesmdSocket = "/var/run/aesmd/aesm.socket"
+)
+
+func dialAesmd() (*net.UnixConn, error) {
+	addr, err := net.ResolveUnixAddr("unix", aesmdSocket)
+	if err != nil {
+		return nil, err
+	}
+
+	conn, err := net.DialUnix("unix", nil, addr)
+	if err != nil {
+		return nil, err
+	}
+
+	return conn, nil
+}
+
+func transmitAesmd(conn *net.UnixConn, req *pb.AesmServiceRequest) ([]byte, error) {
+	rdata, err := proto.Marshal(req)
+	if err != nil {
+		return nil, err
+	}
+
+	msgSize := uint32(len(rdata))
+	byteBuf := bytes.NewBuffer([]byte{})
+	binary.Write(byteBuf, binary.LittleEndian, &msgSize)
+	if _, err = conn.Write(byteBuf.Bytes()); err != nil {
+		return nil, err
+	}
+
+	if _, err = conn.Write(rdata); err != nil {
+		return nil, err
+	}
+
+	rdata = append(rdata[:4])
+	if _, err = conn.Read(rdata); err != nil {
+		return nil, err
+	}
+
+	byteBuf = bytes.NewBuffer(rdata)
+	if err = binary.Read(byteBuf, binary.LittleEndian, &msgSize); err != nil {
+		return nil, err
+	}
+
+	rdata = make([]byte, msgSize)
+	var msgSizeRead int
+	msgSizeRead, err = conn.Read(rdata)
+	if err != nil {
+		return nil, err
+	}
+
+	if msgSizeRead != int(msgSize) {
+		return nil, fmt.Errorf("invalid response size (returned %d, expected %d)",
+			msgSizeRead, msgSize)
+	}
+
+	return rdata, nil
+}
```

Comment:
Above duplicated codes can be removed if you implement GetDcapTargetInfo as a generic interface.